docs/rfcs/0040-multi-cloud-infra.md

```diff
@@ -0,0 +1,304 @@
+Users configure the following application.properties files.
+
+```text
+db.url=jdbc:postgresql://db:5432/featurehub
+db.username=featurehub
+db.password=featurehub
+db.connections=10
+nats.urls=nats://nats:4222
+dacha1.enabled=false
+dacha2.enabled=true
+```
+
+**PostHog**
+
+Users need to create values.yaml such as follows:
+
+```yaml
+cloud: 'aws'
+ingress:
+  hostname: <your-hostname>
+  nginx:
+      enabled: true
+externalPostgresql:
+  # -- External PostgreSQL service host.
+  postgresqlHost:
+  # -- External PostgreSQL service port.
+  postgresqlPort: 5432
+```
+
+Then, deploying service as follows:
+
+```console
+$ helm repo add posthog https://posthog.github.io/charts-clickhouse/
+$ helm repo update
+$ helm upgrade --install -f values.yaml --timeout 30m --create-namespace --namespace posthog posthog posthog/posthog --wait --wait-for-jobs --debug
+```
+
+https://github.com/PostHog/charts-clickhouse
+
+### Conclusion
+
+Since Bucketeer uses Kubernetes, using YAML file and Helm fits into our cases.
+
+## Self-host Bucketeer
+
+Basically, we'll start support self-host by updating each values.yaml.
+
+In the future, we might make it possible to deploy by updating only one YAML file as follows:
+
+[Prototype of design for YAML file](./utils/sample.yml)
+
+### Controversial topic
+
+#### 1. Which middleware requires its own implementation?
+
+| Currently used middlewares | Require to implement in each cloud                           |
+| -------------------------- | ------------------------------------------------------------ |
+| Cloud Pub/Sub              | Yes                                                          |
+| Cloud SQL                  | No (If we support IAM authentication, we need to use AWS SDK) |
+| BigQuery                   | Yes                                                          |
+| Cloud KMS                  | Yes                                                          |
+| MemoryStore                | No                                                           |
+
+##### 2. How to configure values.yaml?
+
+###### Cloud platform
+
+```yaml
+cloudPlatform: gcp
+```
+
+###### Cloud SQL
+
+```yaml
+relationalDatabase:
+  type: mysql
+  password:
+  port:
+```
+
+##### MemoryStore
+
+```yaml
+inMemoryDataStore:
+  password:
+  port:
+```
+
+###### Cloud KMS
+
+```yaml
+keyManagement:
+  gcp:
+    kms:
+      keyId: 
+  aws:
+    kms:
+      ...
+```
+
+###### Cloud Pub/Sub
+
+```yaml
+messaging:
+  gcp:
+    pubsub:
+      topic:
+      subscription:
+      pullerNumGoroutines: 5
+      pullerMaxOutstandingMessages: "1000"
+      pullerMaxOutstandingBytes: "1000000000"
+```
+
+###### BigQuery
+
+```yaml
+bigData:
+  gcp:
+    bigQuery:
+      dataset:
+```
+
+##### 3. Design of directory 
+
+###### pkg directory
+
+The pkg directory will be as follows:
+
+```
+pkg
```

Comment:
I thought we would put all the storage types in one place, like the sample below.
@masaaania,  which one were you thinking?

```
pkg
├── storage
│   ├── bigdata
│       ├── athena
│       └── bigquery
│   ├── inmemory
│       └── redis
│   └── rdb
│       ├── alloydb
│       ├── aurora
│       ├── mysql
│       └── postgres
├── kms
│   ├── aws
│   └── gcp
├── messaging
│   ├── pubsub
│   ├── sns
│   └── sqs

```

Reply:
I prefer to have a `storages` directory, because it is a bit difficult to see more directories under `pkg`.